### ISSUE TITLE
# [Bug] NCCL GPU mapping warnings cause training hangs with DeepSpeed multi-GPU setup

### ISSUE THREAD

**Issue by kkailaasa** [axolotl-ai-cloud/axolotl#2841]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

## Expected Behavior
1. **Dataset processing completes** ✅ (this works)
2. **Model loading should start** across all 4 GPUs with proper memory distribution
3. **FSDP/DeepSpeed initialization** should complete without NCCL warnings
4. **Training loop should begin** with logs showing:
   ```
   Loading checkpoint shards: 100%|██████████| 4/4 [00:XX<00:00, X.XXit/s]
   [INFO] Model loaded successfully across 4 GPUs
   [INFO] DeepSpeed ZeRO optimization enabled
   Training: 0%|          | 0/XXX [00:00<?, ?it/s]
   ```
5. **GPU memory usage** should show:
   - Model sharded across GPUs (~10-15GB per GPU for DeepSeek-V2-Lite)
   - Balanced memory distribution
   - Training progressing normally
6. **No NCCL device mapping warnings** - proper rank-to-GPU assignment should be automatic

## Current Behavior
1. **Dataset processing completes successfully** (52,002 examples processed)
2. **NCCL warnings appear** for all ranks (0,1,2,3)
3. **Training hangs indefinitely** at GPU barrier synchronization
4. **No model loading progress** despite GPUs being active

### 1. Hardware Setup
- 4x NVIDIA A40 GPUs (48GB each)
- Single machine multi-GPU training

### 2. Accelerate Configuration
```bash
accelerate config
# Configuration:
# - Multi-GPU: Yes
# - Number of processes: 4
# - GPU IDs: 0,1,2,3
# - DeepSpeed: No (using axolotl's DeepSpeed config)
# - Mixed precision: bf16
```

### 3. Axolotl Configuration
```yaml
base_model: deepseek-ai/DeepSeek-V2-Lite
trust_remote_code: true
load_in_8bit: true

# DeepSpeed configuration
deepspeed: deepspeed_configs/zero1.json

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca

adapter: lora
lora_r: 8
lora_alpha: 16
gradient_accumulation_steps: 2
micro_batch_size: 4
```

### 4. Command Used
```bash
axolotl train config.yaml
```

### Config yaml

```yaml
base_model: deepseek-ai/DeepSeek-V2-Lite
trust_remote_code: true
load_in_8bit: false
load_in_4bit: false

# DeepSpeed configuration (recommended for multi-GPU)
deepspeed: deepspeed_configs/zero1.json

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca

dataset_prepared_path: last_run_prepared
val_set_size: 0.0
output_dir: ./outputs/mymodel
sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true

# Normal LoRA configuration
adapter: lora
lora_r: 8
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
  - q_proj
  - v_proj
  - k_proj
  - o_proj
  - gate_proj
  - down_proj
  - up_proj
lora_target_linear: true
peft_use_rslora: true

# Training parameters optimized for DeepSpeed
gradient_accumulation_steps: 2
micro_batch_size: 4
num_epochs: 1.0
optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 0.0002
bf16: auto
tf32: false
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false

# Logging and evaluation
logging_steps: 1
flash_attention: true
warmup_steps: 100
evals_per_epoch: 2
saves_per_epoch: 1
weight_decay: 0.0

# Additional settings
save_safetensors: true
strict: false
train_on_inputs: false

# Remove FSDP settings (using DeepSpeed instead)
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey, sorry for late reply. Did you also have the actual error and stack trace to share?

Could you delete the accelerate config file and try again? For single node, we usually don't recommend calling `accelerate config` and specify it within the yaml instead.

**Comment by kkailaasa** [comment]
Oh ok, thanks for the tip - I have since found a solution and was able to fine-tune Deepseek v2-lite successfully, thank you

**Comment by NanoCode012** [comment]
@kkailaasa , could you please share what worked for you for any future readers?

**Comment by kkailaasa** [comment]
> [@kkailaasa](https://github.com/kkailaasa) , could you please share what worked for you for any future readers?

So not exactly sure which was the issue, but bringing the gpu's down to 2 solved it for some reason - it could be something I messed up and that cache did not get cleared, so when I spined up a new node with just 2 it worked.. not a definitive solution unfortunately... sorry 😅
